optimize looping over NOC data movement ops

vroubtsovTT opened the pull request:
This is an onboarding ticket that resolves #496 and that has got a little out of hand. 

The salient changes are:

1. `memref` is now a supported dialect to use in TTKernel IR. There are builtin conversions for {alloca, global, get_global, load, store}.
2. `ConvertTTKernelToEmitCPass` is now a module pass -- this allows us to emit global constants and helper funcs at the module scope level
3. `TTKernelToEmitC` API has been cleaned up a little, to remove methods with odd Region parameters created by caller but populated by callee, etc
4. the original implementation of reduce ops used an "unrolled" loop over operations and could thus emit riscv code that was O(core grid size) in size, with some address calculations repeated redundantly -- the new implementation retains an emitc-level for-loop instead, resulting in much more compact riscv binary
5. a new TTKernel op `TTKernel_NocOpsTablOp` allows defining "arrays of NOC ops" as emitc.globals with memref value types.
6. ttmetal's simple_reduce.mlir test was not included in fb compilation tests, this has been fixed

nsmithtt commented on include/ttmlir/Dialect/TTKernel/IR/TTKernelOps.td:
```diff
@@ -647,6 +647,23 @@ def TTKernel_NocSemaphoreSetMulticastLoopbackOp : TTKernel_Op<"noc_semaphore_set
                          BoolAttr:$multicast_path_reserve);
 }
 
+def TTKernel_NocOpsTableOp : TTKernel_Op<"noc_ops_table"> {
```
NocOps might be a little vague.  Maybe we could call it, `noc_address_table` or `noc_transactions_table` or something to the effect that describes the data it holds.

vroubtsovTT replied:
renamed to slightly-less-vague `noc_transactions_table`

nsmithtt commented on lib/Conversion/TTKernelToEmitC/TTKernelToEmitC.cpp:
```diff
@@ -580,18 +745,12 @@ LogicalResult emitOpRegionAsCpp(Region *region, llvm::raw_ostream &os,
-
-  if (convertTTKernelRegionToEmitC(builder, region, threadType).failed()) {
+  std::optional<mlir::ModuleOp> moduleOp;
+  if (convertTTKernelRegionToEmitC(region, moduleOp, threadType).failed()) {
```
I'm perfectly fine with the way this is written, but just as an FYI, check out `FailureOr` which is a `std::optional` wrapper and is implicitly convertible to `LogicalResult`:

https://github.com/llvm/llvm-project/blob/98d6dd39887361ff2161401614da0b7854234419/llvm/include/llvm/Support/LogicalResult.h#L76

It could enable you to write this without having to use an out param:
```c++
FailureOr<mlir::ModuleOp> failOrModuleOp = convertTTKernelRegionToEmitC(....)
if (!failOrModuleOp)
  return failOrModuleOp;
```

vroubtsovTT replied:
Using FailureOr was more convenient, switched to that.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp:
```diff
@@ -305,37 +263,125 @@ class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
+    llvm::SmallVector<int32_t, (8 * entrySize)> entries;
+    for (auto const &tx : transactions) {
+      assert(tx.type ==
+             type); // all transactions are of the same read/write type
```
Today this might be true, but this will not always be true.  We can handle in a later change when we get there, but maybe keep in mind some ways we could organize the address table by tx type.

vroubtsovTT replied:
This was one assert from the original code that I _didn't_ forget to carry through:
```
    NocTx::Type type = transactions.front().type;
    for (auto tx : transactions) {
      assert(tx.type == type);
```
I don't know eventual future yet and so keeping the constraint (which is easy to remove) seemed the safest choice for now.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp:
```diff
@@ -305,37 +263,125 @@ class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
       std::int64_t addressAlignment, Value *inputCB = nullptr) {
 
+    using std::int32_t;
+    using std::int64_t;
```
I'm not sure where, but LLVM definitely does this somewhere in the header tree, or maybe it just sucks in stdint.h because you can definitely just write `int32_t` and `int64_t`, we do this in many places.

vroubtsovTT replied:
I couldn't find this anywhere in the LLVM project (but that doesn't mean it's not there). I normally deal with this by using a project-wide header that does something like
```
namespace mlir::tt {
  using std::int32_t;
  using std::int64_t;
  ...
}
```
and is included throughout the rest of the project as some "types.h" utility header. Then all our code can benefit from safe short names without exporting them into namespace that aren't ours. Punting on this for now until there's time for small conveniences like that.

nsmithtt commented on lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp:
```diff
@@ -305,37 +263,125 @@ class TTIRToTTMetalLayoutRewriter : public OpRewritePattern<ttir::ToLayoutOp> {
+    for (auto const &tx : transactions) {
+      assert(tx.type ==
+             type); // all transactions are of the same read/write type
+
```
I think we may have lost these assertions:

```c++
assert(nocTx.srcOffset % addressAlignment == 0);
assert(nocTx.dstOffset % addressAlignment == 0);
assert(nocTx.size % addressAlignment == 0);
```

vroubtsovTT replied:
Thank you for spotting this. These have been restored.

nsmithtt reviewed (commented):
Vlad, this is great! Minor comments

vroubtsovTT commented:
Besides addressing the PR comments, that latest commit makes changes to "compact" the noc_transactions_table a little: the x,y coord pair can share the same int32_t entry slot and the last entry elements, numElements, isn't always needed. This works around the "region `NCRISC_LOCAL_DATA_MEM' overflowed" crash in `to_layout.mlir` test. 

This fix is temporary, a better long term solution is needed if we want to emit significant static data structures to support kernels.

nsmithtt reviewed (approved):
Looks great! Thanks Vlad!
